### ISSUE TITLE
Predict using C++

### ISSUE THREAD

**Issue by BenChenCh** [tum-vision/fusenet#8]
Hello,
  When I use trained caffe model to predict with the c++ code changed according to the "cpp_classification" example in Caffe project, the result of this program is "blob<float>",and the values are very small, kline 0.0009; Are there some lines need to be corrected? 
Thanks.
 
`#include <caffe/caffe.hpp>
#ifdef USE_OPENCV
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#endif  // USE_OPENCV
#include <algorithm>
#include <iosfwd>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#ifdef USE_OPENCV
using namespace caffe;  // NOLINT(build/namespaces)
using std::string;

class MyClassifier {
public:
    MyClassifier(const string& model_file,
             const string& trained_file,
             const string& mean_file);
    cv::Mat Predict(const cv::Mat& rgbImg, const cv::Mat& depthImg);
private:
    void SetMean(const string& mean_file);
    void WrapInputLayer(std::vector<cv::Mat>* input_channels);
    void Preprocess(const cv::Mat& rgbImg, const cv::Mat& depthImg,std::vector<cv::Mat>* input_channels);

private:
    shared_ptr<Net<float> > net_;
    cv::Size input_geometry_rgb;
    cv::Size input_geometry_depth;
    int num_channels_rgb;
    int num_channels_depth;
    cv::Mat mean_rgb;
    cv::Mat mean_depth;
};

MyClassifier::MyClassifier(const string& model_file,
                       const string& trained_file,
                       const string& mean_file) {
#ifdef CPU_ONLY
  Caffe::set_mode(Caffe::CPU);
#else
  Caffe::set_mode(Caffe::GPU);
#endif

    /* Load the network. */
    net_.reset(new Net<float>(model_file, TEST));
    net_->CopyTrainedLayersFrom(trained_file);

    CHECK_EQ(net_->num_inputs(), 2) << "Network should have exactly two input.";
    CHECK_EQ(net_->num_outputs(), 1) << "Network should have exactly one output.";
    Blob<float>* input_layer_rgb = net_->input_blobs()[0];
    Blob<float>* input_layer_depth = net_->input_blobs()[1];
    num_channels_rgb = input_layer_rgb->channels();
    num_channels_depth = input_layer_depth->channels();
    CHECK(num_channels_rgb == 3)
        << "RGB Input layer should have 3 channels.";
    CHECK(num_channels_depth == 1)
    << "Depth Input layer should have 1 channels.";
    input_geometry_rgb = cv::Size(input_layer_rgb->width(), input_layer_rgb->height());
    input_geometry_depth = cv::Size(input_layer_depth->width(), input_layer_depth->height());

  /* Load the binaryproto mean file. */
  SetMean(mean_file);
}

/* Load the mean file in binaryproto format. */
void MyClassifier::SetMean(const string& mean_file) {
    BlobProto blob_proto;
    ReadProtoFromBinaryFileOrDie(mean_file.c_str(), &blob_proto);

    /* Convert from BlobProto to Blob<float> */
    Blob<float> mean_blob;
    mean_blob.FromProto(blob_proto);
    //mean prototxt include 1label channel: shoudl subtact 1
    CHECK_EQ(mean_blob.channels()-1, (num_channels_rgb+num_channels_depth))
        << "Number of channels of mean file doesn't match input layer.";

    /* The format of the mean file is planar cv_8uc4 float BGR + grayscale. */
    std::vector<cv::Mat> channels_rgb;
    std::vector<cv::Mat> channels_depth;
    float* data = mean_blob.mutable_cpu_data();
    for (int i = 0; i < (num_channels_rgb+num_channels_depth); ++i) {
        /* Extract an individual channel. */
        cv::Mat channel(mean_blob.height(), mean_blob.width(), CV_8UC1, data);
        if(i<num_channels_rgb) {
            channels_rgb.push_back(channel);
        }else {
            channels_depth.push_back(channel);
        }
        data += mean_blob.height() * mean_blob.width();
    }

    /* Merge the separate channels into a single image. */
    cv::Mat meanRgb,meanDepth;
    cv::merge(channels_rgb, meanRgb);
    cv::merge(channels_depth, meanDepth);

    /* Compute the global mean pixel value and create a mean image
    * filled with this value. */
    cv::Scalar channel_mean_rgb = cv::mean(meanRgb);
    cv::Scalar channel_mean_depth = cv::mean(meanDepth);
    mean_rgb = cv::Mat(input_geometry_rgb, meanRgb.type(), channel_mean_rgb);
    mean_depth = cv::Mat(input_geometry_depth, meanDepth.type(), channel_mean_depth);
}

cv::Mat MyClassifier::Predict(const cv::Mat& rgbImg, const cv::Mat& depthImg) {
    Blob<float>* input_layer_rgb = net_->input_blobs()[0];
    Blob<float>* input_layer_depth = net_->input_blobs()[1];

    input_layer_rgb->Reshape(1, num_channels_rgb,
                       input_geometry_rgb.height, input_geometry_rgb.width);
    input_layer_depth->Reshape(1, num_channels_depth,
                             input_geometry_depth.height, input_geometry_depth.width);
    /* Forward dimension change to all layers. */
    net_->Reshape();

    std::vector<cv::Mat> input_channels;
    WrapInputLayer(&input_channels);

    Preprocess(rgbImg,depthImg, &input_channels);

    net_->Forward();

    /* Copy the output layer to a std::vector */
    Blob<float>* output_layer = net_->output_blobs()[0];
    const float * result=output_layer->cpu_data();
    cv::Mat resultMat(output_layer->height(),output_layer->width(), CV_32FC1,(float *)result);
    cv::Mat resultMatUC;
    return resultMat;
}

void MyClassifier::WrapInputLayer(std::vector<cv::Mat>* input_channels) {
    Blob<float>* input_layer_rgb = net_->input_blobs()[0];
    Blob<float>* input_layer_depth = net_->input_blobs()[1];

    int width = input_layer_rgb->width();
    int height = input_layer_rgb->height();
    float* input_data = input_layer_rgb->mutable_cpu_data();
    for (int i = 0; i < input_layer_rgb->channels(); ++i) {
        cv::Mat channel(height, width, CV_8UC3, input_data);
        input_channels->push_back(channel);
        input_data += width * height;
    }
    width = input_layer_depth->width();
    height = input_layer_depth->height();
    input_data = input_layer_depth->mutable_cpu_data();
    for (int i = 0; i < input_layer_depth->channels(); ++i) {
        cv::Mat channel(height, width, CV_8UC1, input_data);
        input_channels->push_back(channel);
        input_data += width * height;
    }
}

void MyClassifier::Preprocess(const cv::Mat& rgbImg, const cv::Mat& depthImg,
                            std::vector<cv::Mat>* input_channels) {
    /* Convert the input image to the input image format of the network. */
    cv::Mat sample_rgb,sample_depth;
    if (rgbImg.channels() == 4)
        cv::cvtColor(rgbImg, sample_rgb, cv::COLOR_BGRA2BGR);
    else if (rgbImg.channels() == 1 )
        cv::cvtColor(rgbImg, sample_rgb, cv::COLOR_GRAY2BGR);
    else
        sample_rgb = rgbImg;
    //depth image
    if (depthImg.channels() == 4)
        cv::cvtColor(depthImg, sample_depth, cv::COLOR_BGRA2GRAY);
    else if (depthImg.channels() == 3 )
        cv::cvtColor(depthImg, sample_depth, cv::COLOR_BGR2GRAY);
    else if (depthImg.channels() == 1){
        if(depthImg.type()!=CV_8UC1){
            depthImg.convertTo(sample_depth,CV_8UC1,1.0/255);
        }else {
            sample_depth = depthImg;
        }
    }

    cv::Mat sample_resized_rgb;
    if (sample_rgb.size() != input_geometry_rgb)
        cv::resize(sample_rgb, sample_resized_rgb, input_geometry_rgb);
    else
        sample_resized_rgb = sample_rgb;

    cv::Mat sample_resized_depth;
    if (sample_depth.size() != input_geometry_depth)
        cv::resize(sample_depth, sample_resized_depth, input_geometry_depth);
    else
        sample_resized_depth = sample_depth;

    cv::Mat sample_normalized_rgb,sample_normalized_depth;
    cv::subtract(sample_resized_rgb, mean_rgb, sample_normalized_rgb);
    cv::subtract(sample_resized_depth, mean_depth, sample_normalized_depth);

    /* This operation will write the separate BGR planes directly to the
    * input layer of the network because it is wrapped by the cv::Mat
    * objects in input_channels. */
    cv::split(sample_normalized_rgb, *input_channels);
    //depth image is 1 channel
    input_channels->push_back(sample_resized_depth);
}

int main(int argc, char** argv) {
//    if (argc != 6) {
//        std::cerr << "Usage: " << argv[0]
//                  << " deploy.prototxt network.caffemodel"
//                  << " mean.binaryproto img.jpg depth.png" << std::endl;
//        return 1;
//    }

    ::google::InitGoogleLogging(argv[0]);

//    string model_file   = argv[1];//deploy.prototxt
//    string trained_file = argv[2];//caffemodel
//    string mean_file    = argv[3];//mean.prototxt
//    string file = argv[4];
//    string fileDepth = argv[5];//mean.prototxt

    string model_file="segmentation/nyu40-sf1/deploy.prototxt";
    string trained_file="caffemodels_iter_75000.caffemodel";
    string mean_file="db/mean/mean.prototxt";
    string file="NYU0015/fullres/NYU0015.jpg";
    string fileDepth="NYU0015/fullres/NYU0015.png";

    MyClassifier classifier(model_file, trained_file, mean_file);

    std::cout << "---------- Prediction for "
            << file << " ----------" << std::endl;

    cv::Mat img = cv::imread(file, -1);
    cv::Mat imgDepth = cv::imread(fileDepth, -1);
    CHECK(!img.empty()) << "Unable to decode image " << file;
    cv::Mat result=classifier.Predict(img,imgDepth);
//    std::cout<<result.type() <<std::endl;
//    std::cout<<result.cols <<std::endl;
//    std::cout<<result.rows <<std::endl;
}
#else
int main(int argc, char** argv) {
  LOG(FATAL) << "This example requires OpenCV; compile with USE_OPENCV.";
}
#endif  // USE_OPENCV
`

**Comment by hazirbas** [comment]
Hi,
Did you use the default accuracy layer by caffe? Because it doesn't compute the segmentation accuracy correct as far as I remember.

**Comment by BenChenCh** [comment]
Thanks for reply. If I use a RGB image and depth image as input, the output of this net is an object  named "Blob". Here is the code to convert the object to an unsigned char mat:

`    
   //step 1
   std::vector<float> output = Predict(rgbImg,depthImg,rows,cols,channel);

   //step 2
    std::vector<uchar> resultVector;
    int vectorSize=rows*cols;
    for(int i=0;i<vectorSize;i++){
        //step 3
        uchar label=Argmax(output.begin()+i*channel,output.begin()+(i+1)*channel);
        resultVector.push_back(label);
    }
     //step 4
    cv::Mat resultMat(rows,cols, CV_8UC1);
    std::memcpy(resultMat.data, resultVector.data(), resultVector.size());`
 Step 1 : output of the  net: 240*320*41 ( I used 40 classes, so there are 41 probabilities  the for each pixel )
Step 2: create the label mat
Step 3: get the max probability in every 41 element, the index of which is the label of the pixel
Step 4: create an mat filled with the label data

Is the mat in Step 4  a semantic image?

And if I need to get an image like this, what should I do ? Thanks a lot.
![image](https://user-images.githubusercontent.com/12813396/31602186-d32cf2ba-b28e-11e7-908f-6504a64865a8.png)

**Comment by hazirbas** [comment]
Output is a label map (integers) and you can use colormap image, which is not possible with OpenCV. In python you can use PIL.Image for this purpose.

**Comment by BenChenCh** [comment]
I am stuck now on using caffemodel to do semantic segmentation. I have tried for several days. Could you please share the python script to do so?

**Comment by hazirbas** [comment]
Hi,
I can share the following code snippet. Once you forward the image through the network, you can compute the prediction as:

`prediction   = np.argmax( output['segmentation-blob'][batch_id], axis = 0).astype( np.uint8 )`

**Comment by BenChenCh** [comment]
Hi，
     Thanks for your reply. Does it need the mean file or array in python script to do semantic segmentation?

**Comment by hazirbas** [comment]
You forward the images using CAFFE prototxt and images are all processed in the network.

**Comment by BenChenCh** [comment]
Here is my python code to prediction. Is this the right way? If not, could you tell me?
`
import caffe
import numpy as np
from PIL import Image
import matplotlib.pyplot as plt
from scipy import misc

deploy='deploy.prototxt'    #deploy
caffe_model='caffemodels_.caffemodel'   #caffemodel
net = caffe.Net(deploy,caffe_model,caffe.TEST)   # load model and network

transformer = caffe.io.Transformer({'data': net.blobs['rgb'].data.shape}) 
transformer.set_transpose('data', (2,0,1))    #change dimens: eg from (28,28,3) to (3,28,28)

transformerDepth = caffe.io.Transformer({'data': net.blobs['depth'].data.shape}) 
transformerDepth.set_transpose('data', (2,0,1)) 


img='rgb.jpg'    #image path
depthImg='depth.png'

imgRgb=caffe.io.load_image(img)#rgb image
depthImg=caffe.io.load_image(depthImg,False) #depth image 
net.blobs['rgb'].data[...] = transformer.preprocess('data',imgRgb)
net.blobs['depth'].data[...] = transformerDepth.preprocess('data',depthImg)

#test
out = net.forward()
prob= net.blobs['prob'].data 

prediction = np.argmax(prob[0], axis = 0).astype( np.uint8 )
plt.imshow(prediction)
plt.show()
`

**Comment by hazirbas** [comment]
prediction is integer array, you plot it but you should use coloring to see the final result. Please look into [http://pillow.readthedocs.io/en/3.4.x/reference/ImagePalette.html](PIL ImagePalette) to colorize the predictions:

**Comment by BenChenCh** [comment]
the result of prediction in my experience is bad. 
Original image: 
![nyu0095](https://user-images.githubusercontent.com/12813396/31808742-0f8911ec-b5a7-11e7-9af9-09cf2cf95603.jpg)
Prediction image:
![prediction95](https://user-images.githubusercontent.com/12813396/31808764-2449d85a-b5a7-11e7-8de1-9e267bc84ed3.png)

So I thought I created the lmdb file in a wrong way. 
According to the slice layer, i guess there were 5 channels for each element: 3 channels for  rgb , 1 channels for depth , 1 channel for label . When creating lmdb file , I fused a rgb image,  a depth image and a label image into a image with 5 channels and the same size as theirs . Is that right?
`layer {
  name: "slice_data"
  type: "Slice"
  bottom: "data"
  top: "rgb"
  top: "depth"
  top: "label"
  slice_param {
    slice_dim: 1
    slice_point: 3
    slice_point: 4
  }
}`
